homeassistant/components/unifi/sensor.py

```diff
@@ -192,6 +199,85 @@ def async_device_state_value_fn(hub: UnifiHub, device: Device) -> str:
     return DEVICE_STATES[device.state]
 
 
+@callback
+def async_client_wan_monitor_supported_fn(
+    monitor_target: str, hub: UnifiHub, obj_id: str
+) -> bool:
+    """Determine if a client supports WAN monitoring."""
+    device = hub.api.devices[obj_id]
+
+    uptime_stats = device.uptime_stats
+    if uptime_stats is None:
+        return False
+
+    wan = uptime_stats["WAN"]
+
+    monitors = wan["monitors"]
+
+    return any(monitor_target in item["target"] for item in monitors)
+
+
+@callback
+def async_client_wan_monitor_latency(
+    wan: Literal["WAN", "WAN2"],
+    monitor_target: str,
+    hub: UnifiHub,
+    device: Device | None,
+) -> int | None:
+    """Retrieve the monitor target from WAN monitors."""
+
+    if device is None or device.uptime_stats is None:
+        return None
+
+    uptime_stats = device.uptime_stats
+
+    if uptime_stats_wan := uptime_stats.get(wan):
+        for item in uptime_stats_wan["monitors"]:
+            if monitor_target in item["target"]:
+                return item.get("latency_average", 0)
+
+    return None
+
+
+@callback
+def make_wan_latency_entity_description(
+    wan: Literal["WAN", "WAN2"], name: str, monitor_target: str
+) -> UnifiSensorEntityDescription:
+    """Create a WAN latency sensor entity description."""
+
+    return UnifiSensorEntityDescription[Devices, Device](
+        key=f"{name} WAN latency",
+        entity_category=EntityCategory.DIAGNOSTIC,
+        native_unit_of_measurement=UnitOfTime.MILLISECONDS,
+        state_class=SensorStateClass.MEASUREMENT,
+        device_class=SensorDeviceClass.DURATION,
+        entity_registry_enabled_default=False,
+        api_handler_fn=lambda api: api.devices,
+        available_fn=async_device_available_fn,
+        device_info_fn=async_device_device_info_fn,
+        name_fn=lambda _: f"{name} WAN latency",
+        object_fn=lambda api, obj_id: api.devices[obj_id],
+        supported_fn=lambda hub, obj_id: async_client_wan_monitor_latency(
+            wan, monitor_target, hub, hub.api.devices[obj_id]
+        )
+        is not None,
+        unique_id_fn=lambda hub, obj_id: f"{name.lower}_wan_latency-{obj_id}",
+        value_fn=partial(async_client_wan_monitor_latency, wan, monitor_target),
+    )
+
+
+def make_wan_latency_sensors(
+    wan: Literal["WAN", "WAN2"],
+) -> tuple[UnifiSensorEntityDescription, ...]:
+    """Create WAN latency sensors from WAN monitor data."""
+
+    return (
+        make_wan_latency_entity_description(wan, "Microsoft", "microsoft"),
+        make_wan_latency_entity_description(wan, "Google", "google"),
+        make_wan_latency_entity_description(wan, "Cloudflare", "1.1.1.1"),
+    )
```

Comment:
You could probably change this to return the full dict of entity descriptions and not need to input WAN/WAN2

Reply:
I have a little hard time do figure this out. 
```python
    descriptions = []
    for wan in ("WAN", "WAN2"):
        descriptions += [
            make_wan_latency_entity_description(wan, "Microsoft", "microsoft"),
            make_wan_latency_entity_description(wan, "Google", "google"),
            make_wan_latency_entity_description(wan, "Cloudflare", "1.1.1.1"),
        ]

    return tuple(descriptions)
```

Will not work, as `make_wan_latency_entity_description`'s wan is a `Literal["WAN", "WAN2"]`
So mypy will complain saying `Argument 1 to "make_wan_latency_entity_description" has incompatible type "str"; expected "Literal['WAN', 'WAN2']"  [arg-type]`

So is there a way to either loop over `Literal["WAN", "WAN2"]` or fix it somehow else?

Reply:
Pushed the code to see :)